design/general-progress-monitoring.md

```diff
@@ -0,0 +1,496 @@
+                                                // For a disk, an item would be a byte and itemsToComplete would be the
+                                                // total size to transfer (may be less than the size of a volume if
+                                                // performing an incremental) and itemsCompleted is the number of bytes
+                                                // transferred.  On successful completion, itemsCompleted and itemsToComplete
+                                                // should be the same
+        started, updated time.Time              // When the upload was started and when the last update was seen.  Not all
+                                                // systems retain when the upload was begun, return Time 0 (time.Unix(0, 0))
+                                                // if unknown.
+    }
+
+### VolumeSnapshotter changes
+
+Two new methods will be added to the VolumeSnapshotter interface:
+
+    Progress(snapshotID string) (UploadProgress, error)
+    Cancel(snapshotID string) (UploadProgress, error)
+
+Open question: Does VolumeSnapshotter need Cancel, or is that only needed for BIA/RIA?
+
+Progress will report the current status of a snapshot upload.  This should be callable at
+any time after the snapshot has been taken.  In the event a plug-in is restarted, if the operationID
+(snapshot ID) continues to be valid it should be possible to retrieve the progress.
+
+`error` is set if there is an issue retrieving progress.  If the snapshot is has encountered an
+error during the upload, the error should be returned in OperationProgress and error should be nil.
+
+### BackupItemAction and RestoreItemAction plug-in changes
+
+Currently CSI snapshots and the Velero Plug-in for vSphere are implemented as BackupItemAction
+plugins.  While the majority of BackupItemAction plugins do not take snapshots or upload data, this
+functionality is useful for any longstanding plugin operation managed by an external
+process/controller so we will modify BackupItemAction and RestoreItemAction to optionally return an
+operationID in addition to the modified item.
+
+The Backup/RestoreItemAction plugin identifier as well as the Item and Snapshot ID will be stored in
+the `<backup-name>-itemoperations.json.gz`.  When checking for progress, this info will be used to
+select the appropriate Backup/RestoreItemAction plugin to query for progress.
+
+Velero can attempt to cancel an operation by calling the Cancel API call on the BIA/RIA. The plugin
+can then take any appropriate action as needed. Cancel will be called on backup deletion, and
+possibly upon reaching timeouts, if timeout support is included in this feature.
+
+_NotApplicable_ should only be returned if the Backup/RestoreItemAction plugin should not be
+handling the item.  If the SnapshotItemAction plugin should handle the item but, for example, the
```

Comment:
Fixed, also updated wording to reflect that Cancel/Progress will return an error rather than a populated OperationProgress struct if the operation doesn't exist.